flypipe/run_context.py

```diff
@@ -0,0 +1,73 @@
+from pyspark.sql import SparkSession
+
+from flypipe.cache import CacheMode
+from flypipe.cache.cache_context import CacheContext
+from flypipe.config import get_config, RunMode
+
+
+class RunContext:  # pylint: disable=too-few-public-methods
+    """
+    RunContext is a model held by each run that holds run information that is tied to a particular run,
+    such as parameters, cache mode and inputs.
+    """
+
+    def __init__(self,
+                 spark: SparkSession = None,
+                 parallel: bool = None,
+                 provided_inputs: dict = None,
+                 pandas_on_spark_use_pandas: bool = False,
+                 parameters: dict = None,
+                 cache_modes: dict = None):
+
+        self._spark = spark
+        self._parallel = get_config("default_run_mode") == RunMode.PARALLEL.value if parallel is None else parallel
+        self._provided_inputs = provided_inputs or {}
+        self._pandas_on_spark_use_pandas = False if pandas_on_spark_use_pandas is None else pandas_on_spark_use_pandas
+        self._parameters = parameters or {}
+        self._cache_modes = cache_modes or {}
+
+    def copy(self):
+        return RunContext(
```

Comment:
Ah yes, deep copy on an object will deep copy all it's fields, which as you found out for spark will break on singleton objects. File references aka open(file, 'r') are another thing I think it will break on.

Reply:
cahnged to 
```
    def copy(self):
        return copy(self)

```